src/main/java/gregtech/api/capability/impl/CleanroomLogic.java

```diff
@@ -0,0 +1,228 @@
+        if (hasEnoughPower()) amountToClean = BASE_CLEAN_AMOUNT * (getTierDifference() + 1);
+        else amountToClean = -BASE_CLEAN_AMOUNT;
+
+        // each maintenance problem lowers gain by 1 and increases loss by 1
+        if (hasMaintenance) amountToClean -= ((IMaintenance) metaTileEntity).getNumMaintenanceProblems();
+        ((ICleanroomProvider) metaTileEntity).adjustCleanAmount(amountToClean);
+    }
+
+    protected boolean consumeEnergy(boolean simulate) {
+        return ((ICleanroomProvider) metaTileEntity).drainEnergy(simulate);
+    }
+
+    /**
+     * @return true if the cleanroom is able to drain energy, else false
+     */
+    protected boolean hasEnoughPower() {
+        if (!consumeEnergy(true)) {
+            if (progressTime >= 2) {
+                if (ConfigHolder.machines.recipeProgressLowEnergy) this.progressTime = 1;
+                else this.progressTime = Math.max(1, progressTime - 2);
+
+                hasNotEnoughEnergy = true;
+            }
+            return false;
+        }
+
+        if (this.hasNotEnoughEnergy && ((ICleanroomProvider) metaTileEntity).getEnergyInputPerSecond() > 19L * GTValues.VA[((ICleanroomProvider) metaTileEntity).getEnergyTier()]) {
+            this.hasNotEnoughEnergy = false;
+        }
+
+        return true;
+    }
+
+    /**
+     * @return true if the cleanroom is active
+     */
+    public boolean isActive() {
+        return this.isActive;
+    }
+
+    /**
+     * @param active the new state of the cleanroom's activity: true to change to active, else false
+     */
+    public void setActive(boolean active) {
+        if (this.isActive != active) {
+            this.isActive = active;
+            this.metaTileEntity.markDirty();
+            if (metaTileEntity.getWorld() != null && !metaTileEntity.getWorld().isRemote) {
+                this.metaTileEntity.writeCustomData(GregtechDataCodes.IS_WORKING, buf -> buf.writeBoolean(active));
+            }
+        }
+    }
+
+    /**
+     * @param workingEnabled the new state of the cleanroom's ability to work: true to change to enabled, else false
+     */
+    public void setWorkingEnabled(boolean workingEnabled) {
+        this.isWorkingEnabled = workingEnabled;
+        metaTileEntity.markDirty();
+    }
+
+    /**
+     * @return whether working is enabled for the logic
+     */
+    public boolean isWorkingEnabled() {
+        return isWorkingEnabled;
+    }
+
+    /**
+     * @return whether the cleanroom is currently working
+     */
+    public boolean isWorking() {
+        return isActive && !hasNotEnoughEnergy && isWorkingEnabled;
+    }
+
+    /**
+     * @return the current progress towards completing one cycle of the cleanroom
+     */
+    public int getProgressTime() {
+        return this.progressTime;
+    }
+
+    public int getMaxProgress() {
+        return MAX_PROGRESS;
+    }
+
+    public double getProgressPercent() {
+        return getProgressTime() * 1.0 / getMaxProgress();
+    }
```

Comment:
Why is this a double when both parameters are int, and the only place that uses this method casts it back to an int as well?